### PR TITLE
[ENG-3035] Update Registration Cards

### PR DESCRIPTION
**Pull request by chth0n1x**
-   Ticket: https://openscience.atlassian.net/browse/ENG-3035
-   Feature flag: n/a

## Purpose

The purpose of this ticket was to add an update button to enable changes for versioning.

### CODE REVIEW

**Review by futa-ikeda — CHANGES_REQUESTED** [review]
Had some feedback and a couple of questions

File: app/models/registration.ts [hunk 1]
```diff
@@ -104,10 +104,12 @@ export default class RegistrationModel extends NodeModel.extend(Validations) {
     @attr('fixstring') reviewsState!: RegistrationReviewStates;
     @attr('fixstring') iaUrl?: string;
     @attr('array') providerSpecificMetadata!: ProviderMetadata[];
-    @attr('fixstring') revisionState!: RevisionReviewStates;
+    @attr('fixstring') revisionState?: RevisionReviewStates;
+    @attr('object') latestSchemaResponse?: SchemaResponseModel;
 
     // Write-only attributes
     @attr('array') includedNodeIds?: string[];
+    @attr('boolean') createDoi?: boolean;
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply app/models/registration.ts:112]: Got it, ok I just removed it.

File: app/models/registration.ts [hunk 2]
```diff
@@ -149,6 +151,7 @@ export default class RegistrationModel extends NodeModel.extend(Validations) {
     // Write-only relationships
     @belongsTo('draft-registration', { inverse: null })
     draftRegistration!: DraftRegistrationModel;
+    static reviewsState: string;
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply app/models/registration.ts:154]: Ok, removed!

File: lib/osf-components/addon/components/node-card/component.ts [hunk 3]
```diff
@@ -28,9 +43,65 @@ export default class NodeCard extends Component {
...
+
+    @task
+    @waitFor
+    async getLatestRevision(registration: RegistrationModel) {
+        if (!registration) {
+            const notARegistrationError = this.intl.t('registries.update_dropdown.not_a_registration_error');
+            return this.toast.error(notARegistrationError);
+        }
+
+        if (registration.reviewsState !== 'accepted') {
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/component.ts:62]: Got it, logic now reads: `if (registration.reviewsState !== 'accepted' && registration.reviewsState !== 'embargo') {` on line 62.

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 4]
```diff
@@ -87,13 +103,21 @@
                                 >
                                     {{t 'general.delete'}}
                                 </BsButton>
+                                <BsButton
+                                    data-analytics-scope='Delete'
+                                    local-class='NodeCard__update-link'
+                                    @onClick={{action @delete @node}}
+                                >
+                                    {{t 'general.delete'}}
+                                </BsButton>
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs:111]: Makes sense. Ok, it's removed!

[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs:111]: Actually, would you like the whole menu item from line 97 to 112 deleted including the dropdown link?

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 5]
```diff
@@ -34,14 +34,30 @@
...
+                    <span class='label label-danger'>{{t 'node_card.registration.statuses.revision_in_progress'}}</span> |
+                {{else if (eq @node.revisionState 'unapproved')}}
+                    <span class='label label-danger'>{{t 'node_card.registration.statuses.revision_pending_admin_approval'}}</span> |
+                {{else if (eq @node.revisionState 'in_progress')}}
+                    <span class='label label-danger'>{{t 'node_card.registration.statuses.revision_pending_moderation'}}</span> |
+                {{else if (eq @node.revisionState 'pending_moderation')}}
+                    <span class='label label-danger'>{{t 'node_card.registration.statuses.revision_pending_admin_approval'}}</span> |
+                {{else if (eq @node.revisionState 'approved')}}
+                    <span class='label label-danger'>{{t 'node_card.registration.statuses.revision_pending_moderation'}}</span> |
+                {{/if}}
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs]: I was actually considering this. Would we like to set logic for the Approved updates/edits to show as 'Latest' update the first time the user logs in and sees or clicks on the Registration? 

Also yes the translation strings are updated locally, with changes pushed this morning.

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 6]
```diff
@@ -157,20 +181,46 @@
...
+                            @route='resolve-guid'
+                            @models={{array @node.id}}
+                        >
+                            <Button>
+                                {{t 'node_card.view_button'}}
+                            </Button>
+                        </OsfLink>
+                        {{#if this.shouldShowUpdateButton}}
+                            {{#unless (not-eq this.node.reviewsState 'accepted')}}
+                                {{#if this.latestSchemaResponse}}
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs]: I agree. I just accepted the suggested code edit, thank you. I will begin testing it as well shortly.

File: lib/osf-components/addon/components/node-card/component.ts [hunk 7]
```diff
@@ -28,9 +43,65 @@ export default class NodeCard extends Component {
...
+        await newRevision.save();
+        this.router.transitionTo('registries.edit-revision', newRevision.id);
+    }
+
+    @computed('node.isRegistration')
+    get shouldShowUpdateButton(): boolean {
+        let showUpdateButton = false;
+        if (this.node && this.node.isRegistration) {
+            const registration = this.node as Registration;
+            if (registration.reviewsState === 'accepted' || registration.reviewsState === 'embargo') {
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/component.ts:99]: Ok, making those changes now.

[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/component.ts:99]: Ok, the code is now augmented `       if (!this.node?.userHasAdminPermission){
            return;
        }
        if (registration.reviewsState !== 'accepted' && registration.reviewsState !== 'embargo') {
            if (registration.revisionState !== 'approved') {
                return;
            }
        } else {
            try {
                const revisions = await registration.queryHasMany('schemaResponses');`

I can push changes shortly if that looks correct/best! If not, just let me know and I can update the logic.

File: lib/registries/addon/components/make-decision-dropdown/component.ts [hunk 8]
```diff
@@ -15,7 +15,6 @@ import RegistrationModel,
     RegistrationReviewStates,
     reviewsStateToDecisionMap,
     NonActionableRegistrationStates,
-    ActionableRevisionStates,
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/registries/addon/components/make-decision-dropdown/component.ts:18]: Of course, re-adding the missing import now.

File: lib/osf-components/addon/components/node-card/component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (futa-ikeda) [reply lib/osf-components/addon/components/node-card/component.ts:101]: it's a little hard to read, but I think it looks right. Except the last else block you have, it is fetching revisions, it doesn't need to do that.

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 9]
```diff
@@ -34,14 +34,30 @@
...
+                
+                {{@node.title}}
+                   
                 <OsfLink
                     data-analytics-name='Title'
                     data-test-node-title='{{@node.id}}'
                     @route='resolve-guid'
                     @models={{array @node.id}}
                 >
-                    {{@node.title}}
```
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs:44]: Yes it does. Thank you Futa.

[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs:44]: I've readded the button now.

[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs:44]: I've readded the link back now.

File: lib/osf-components/addon/components/node-card/template.hbs (hunk 6 above)
[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs]: Ok, the logic in place is ` {{else if this.shouldShowLinkToUpdate}}
                <Button
                    {{on 'click' (perform this.createNewSchemaResponse)}}
                    disabled={{this.createNewSchemaResponse.isRunning}}
                />`

[the remark this answers is not in the document]
Comment (chth0n1x) [reply lib/osf-components/addon/components/node-card/template.hbs]: Will this resolve the second condition?

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 10]
```diff
@@ -34,14 +34,29 @@
                 {{#if @node.archiving}}
                     <span class='label label-primary'>{{t 'node_card.registration.statuses.archiving'}}</span> |
                 {{/if}}
+           
+                {{#if @node.revisionState.id}}
+                {{else if (eq @node.revisionState 'unapproved')}}
```
Comment (futa-ikeda) [inline lib/osf-components/addon/components/node-card/template.hbs:39]:
```suggestion
                {{#if (eq @node.revisionState 'unapproved')}}
```

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 11]
```diff
@@ -157,20 +164,51 @@
...
+                            </Button>
+                        </OsfLink>
+                        {{#if this.shouldShowUpdateButton}}
+                            <Button
+                              {{on 'click' (perform this.createNewSchemaResponse)}}
+                              disabled={{this.createNewSchemaResponse.isRunning}}
+                            />
+                          {{else if this.shouldShowLinkToUpdate}}
+                            <OsfLink
+                             ......
```
Comment (futa-ikeda) [inline lib/osf-components/addon/components/node-card/template.hbs:188]:
```suggestion
```

File: lib/osf-components/addon/components/node-card/template.hbs [hunk 12]
```diff
@@ -157,20 +164,51 @@
...
+                                        >
+
+                                        <Button
+                                            local-class='RegistrationCard_update'
+                                        >
+                                            {{t 'node_card.update_button'}}
+                                        </Button>
+                                    </OsfLink>
+                                {{/if}}
+                            {{/unless}}
```
Comment (futa-ikeda) [inline lib/osf-components/addon/components/node-card/template.hbs:203]:
```suggestion
```

File: lib/osf-components/addon/components/node-card/template.hbs (hunk 6 above)
[the remark this answers is not in the document]
Comment (futa-ikeda) [reply lib/osf-components/addon/components/node-card/template.hbs]: I think the template file looks correct right now the way it is. We may want to tweak some of the logic in the `shouldShowUpdateButton` computed property, but we can update those in the component.ts file. We will also need to add a new computed property called `shouldShowLinkToUpdate` in the component.ts file so a link to the existing revision in progress is shown appropriately:

```
                        {{#if this.shouldShowUpdateButton}}
                            <Button
                              {{on 'click' (perform this.createNewSchemaResponse)}}
                              disabled={{this.createNewSchemaResponse.isRunning}}
                            />
                          {{else if this.shouldShowLinkToUpdate}}
                            <OsfLink
                                        data-analytics-name='Update'
                                        data-test-update-button='{{@node.id}}'
                                        @route='registries.edit-revision.index'
                                        @models={{array this.latestSchemaResponse.id}}
                                        >

                                        <Button
                                            local-class='RegistrationCard_update'
                                        >
                                            {{t 'node_card.update_button'}}
                                        </Button>
                                    </OsfLink>
                                {{/if}}
```

File: lib/osf-components/addon/components/node-card/template.hbs (hunk 5 above)
[the remark this answers is not in the document]
Comment (futa-ikeda) [reply lib/osf-components/addon/components/node-card/template.hbs]: > Would we like to set logic for the Approved updates/edits to show as 'Latest' update the first time the user logs in and sees or clicks on the Registration?

We could, but I don't think it is necessary

**Comment by adlius** [comment]
Closed in favor of #1346

**Pull request closed without merging**